In [2]:
import numpy as np
import pandas_profiling
from scipy.stats import mode
import scipy as sc
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import os
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

## **Data Exploration**

In [3]:
#Four diff datasets: Glasses, smartphone, smartwatch and report
# Data can be found at https://www.kaggle.com/sasanj/human-activity-smart-devices
glasses_df = pd.read_csv(os.getcwd()+os.sep+'data'+os.sep+'glasses.csv')
watch_df = pd.read_csv(os.getcwd()+os.sep+'data'+os.sep+'smartwatch.csv')
phone_df = pd.read_csv(os.getcwd()+os.sep+'data'+os.sep+'smartphone.csv')
report_df = pd.read_csv(os.getcwd()+os.sep+'data'+os.sep+'report.csv')
pd.set_option('display.max_columns', None)

In [4]:
glasses_df.sort_values(by=['DATE'],ascending=False, inplace= True)
glasses_df = glasses_df[["DATE","ACC_X","ACC_Y","ACC_Z","GYRO_X","GYRO_Y","GYRO_Z","EOG_L","EOG_R","EOG_H","EOG_V"]]
glasses_df.head(10)

DATE  ACC_X  ACC_Y  ACC_Z  GYRO_X  GYRO_Y  GYRO_Z  \
3975265  2017/07/13 22:20:30.964  -4968   9414 -12622    1052    1176    -154   
3975264  2017/07/13 22:20:30.944  -5082   9531 -12473     881    1062    -189   
3975263  2017/07/13 22:20:30.934  -5170   9579 -12492     828     963    -214   
3975262  2017/07/13 22:20:30.914  -5280   9682 -12471     581     812    -271   
3975261  2017/07/13 22:20:30.894  -5345   9790 -12500     336     638    -365   
3975260  2017/07/13 22:20:30.874  -5375   9829 -12553      16     488    -470   
3975259  2017/07/13 22:20:30.854  -5382   9888 -12615    -330     376    -611   
3975258   2017/07/13 22:20:30.84  -3159   9234 -13234    -326    -693     126   
3975257  2017/07/13 22:20:30.824  -5244   9814 -12379    -884      41    -789   
3975256  2017/07/13 22:20:30.804  -5316   9886 -12501   -1101    -231    -847   

         EOG_L  EOG_R  EOG_H  EOG_V  
3975265   -147      2   -149     72  
3975264   -159     -4   -155     81  
3975263   -176    -19   -157     97  
3975262   -183    -28   -155    105  
3975261   -186    -32   -154    109  
3975260   -174    -20   -154     97  
3975259   -185    -22   -163    103  
3975258   -130     28   -158     51  
3975257   -164     -8   -156     86  
3975256   -170    -21   -149     95

In [5]:
watch_df.sort_values(by=['timestamp'],ascending=False, inplace= True)
watch_df = watch_df[["timestamp","source","values"]]
watch_df.head(10)

timestamp      source         values
200470  2017-07-13 20:21:36.105    pressure   ['989.3108']
200469  2017-07-13 20:21:36.105    pressure   ['989.3108']
200468  2017-07-13 20:21:35.124  heart_rate  ['57.0', '1']
200467  2017-07-13 20:21:35.124  heart_rate  ['57.0', '1']
200466  2017-07-13 20:21:35.060  heart_rate  ['56.0', '1']
200465  2017-07-13 20:21:35.060  heart_rate  ['56.0', '1']
200464  2017-07-13 20:21:35.020  heart_rate  ['55.0', '1']
200463  2017-07-13 20:21:35.020  heart_rate  ['55.0', '1']
200462  2017-07-13 20:21:35.009  heart_rate  ['54.0', '1']
200461  2017-07-13 20:21:35.009  heart_rate  ['54.0', '1']

In [6]:
phone_df.sort_values(by=['timestamp'],ascending=False, inplace= True)
phone_df = phone_df[["timestamp","source","values"]]
phone_df.head(10)

timestamp               source  \
1528217  2017-07-13 20:13:38.399  linear_acceleration   
1528216  2017-07-13 20:13:38.194                 wifi   
1528215  2017-07-13 20:13:37.681         magnetometer   
1528214  2017-07-13 20:13:37.668                audio   
1528213  2017-07-13 20:13:36.668                audio   
1528212  2017-07-13 20:13:36.478          orientation   
1528211  2017-07-13 20:13:35.928              gravity   
1528210  2017-07-13 20:13:35.917             pressure   
1528209  2017-07-13 20:13:35.663                audio   
1528208  2017-07-13 20:13:35.202            gyroscope   

                                                    values  
1528217       ['-0.07805297', '-0.3648016', '0.009873864']  
1528216               [7207, 5500, 8953, 4838, 3544, 3564]  
1528215           ['-47.695847', '14.789566', '11.728146']  
1528214         ['2441.0', '5245.9517', '32767.0', '30.0']  
1528213          ['2046.0', '5246.024', '32767.0', '30.0']  
1528212  ['137.4753975006187', '27.844820019593158', '-...  
1528211            ['8.826892', '-4.260077', '-0.3288707']  
1528210                                      ['987.33563']  
1528209         ['2340.0', '5246.1064', '32767.0', '30.0']  
1528208  ['-0.015442833', '-0.0089864135', '-0.004155917']

In [7]:
report_df.info()
report_df = report_df[["activity_type","duration","from","to"]]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151 entries, 0 to 150
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   index          151 non-null    int64 
 1   activity_type  151 non-null    object
 2   duration       151 non-null    object
 3   from           151 non-null    object
 4   to             151 non-null    object
 5   comment        3 non-null      object
dtypes: int64(1), object(5)
memory usage: 7.2+ KB


In [8]:
# pandas_profiling.ProfileReport(report_df)

### **Data preprocessing**
- Merging smartwatch and phone data
- Casting timestamp as datetime 

In [9]:
#Merging phone and watch data
frames = [watch_df, phone_df]
device_df = pd.concat(frames)
device_df.head(10)

timestamp      source         values
200470  2017-07-13 20:21:36.105    pressure   ['989.3108']
200469  2017-07-13 20:21:36.105    pressure   ['989.3108']
200468  2017-07-13 20:21:35.124  heart_rate  ['57.0', '1']
200467  2017-07-13 20:21:35.124  heart_rate  ['57.0', '1']
200466  2017-07-13 20:21:35.060  heart_rate  ['56.0', '1']
200465  2017-07-13 20:21:35.060  heart_rate  ['56.0', '1']
200464  2017-07-13 20:21:35.020  heart_rate  ['55.0', '1']
200463  2017-07-13 20:21:35.020  heart_rate  ['55.0', '1']
200462  2017-07-13 20:21:35.009  heart_rate  ['54.0', '1']
200461  2017-07-13 20:21:35.009  heart_rate  ['54.0', '1']

In [10]:
device_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1728689 entries, 200470 to 0
Data columns (total 3 columns):
 #   Column     Dtype 
---  ------     ----- 
 0   timestamp  object
 1   source     object
 2   values     object
dtypes: object(3)
memory usage: 52.8+ MB


In [11]:
glasses_df['timestamp'] = glasses_df['DATE']
glasses_df = glasses_df[["timestamp","ACC_X","ACC_Y","ACC_Z","GYRO_X","GYRO_Y","GYRO_Z","EOG_L","EOG_R","EOG_H","EOG_V"]]

In [12]:
device_df['timestamp'] = pd.to_datetime(device_df['timestamp'])

In [13]:
report_df['from'] = pd.to_datetime(report_df['from'])
report_df['to'] = pd.to_datetime(report_df['to'])
report_df.info()
report_df.head(10)
device_df.info()
device_df.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 151 entries, 0 to 150
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   activity_type  151 non-null    object        
 1   duration       151 non-null    object        
 2   from           151 non-null    datetime64[ns]
 3   to             151 non-null    datetime64[ns]
dtypes: datetime64[ns](2), object(2)
memory usage: 4.8+ KB


activity_type duration                from                  to
0   Video games    01:43 2017-07-01 21:34:00 2017-07-01 23:17:00
1   In computer    00:03 2017-07-01 21:29:00 2017-07-01 21:32:00
2       At home    13:35 2017-07-01 21:13:00 2017-07-02 10:49:00
3   In computer    00:05 2017-07-01 21:08:00 2017-07-01 21:13:00
4           Eat    00:18 2017-07-01 20:49:00 2017-07-01 21:07:00
5   In computer    00:20 2017-07-01 18:26:00 2017-07-01 18:46:00
6           Eat    00:13 2017-07-01 16:20:00 2017-07-01 16:33:00
7         Sleep    04:31 2017-07-01 11:00:00 2017-07-01 15:31:00
8           Eat    00:10 2017-07-02 21:55:00 2017-07-02 22:05:00
9   Watching TV    01:59 2017-07-02 20:10:00 2017-07-02 22:10:00

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1728689 entries, 200470 to 0
Data columns (total 3 columns):
 #   Column     Dtype         
---  ------     -----         
 0   timestamp  datetime64[ns]
 1   source     object        
 2   values     object        
dtypes: datetime64[ns](1), object(2)
memory usage: 52.8+ MB


timestamp      source         values
200470 2017-07-13 20:21:36.105    pressure   ['989.3108']
200469 2017-07-13 20:21:36.105    pressure   ['989.3108']
200468 2017-07-13 20:21:35.124  heart_rate  ['57.0', '1']
200467 2017-07-13 20:21:35.124  heart_rate  ['57.0', '1']
200466 2017-07-13 20:21:35.060  heart_rate  ['56.0', '1']
200465 2017-07-13 20:21:35.060  heart_rate  ['56.0', '1']
200464 2017-07-13 20:21:35.020  heart_rate  ['55.0', '1']
200463 2017-07-13 20:21:35.020  heart_rate  ['55.0', '1']
200462 2017-07-13 20:21:35.009  heart_rate  ['54.0', '1']
200461 2017-07-13 20:21:35.009  heart_rate  ['54.0', '1']

### **Data labelling**
Using report data to label activity type with sensor data. Sampling to make data labeling complete in resonable time.

In [16]:
device_df['activity_type'] = ''
sample_df = device_df.sample(frac=0.1,replace=True, random_state=1)
for report_index, report_row in report_df.iterrows():
    for device_index, device_row in sample_df.iterrows():
        if report_row['from'] <= device_row['timestamp'] <= report_row['to']:
            sample_df.loc[device_index, "activity_type"] = report_row['activity_type']

In [17]:
sample_df.head(10)

timestamp         source  \
72433   2017-07-03 19:56:35.104   step_counter   
1236933 2017-07-11 15:37:08.031          audio   
1257764 2017-07-11 17:37:08.718   magnetometer   
1237425 2017-07-11 15:39:58.565        gravity   
892199  2017-07-07 17:55:46.660       pressure   
1357285 2017-07-12 15:26:12.372    orientation   
127121  2017-07-09 08:17:23.866     heart_rate   
562529  2017-07-04 13:33:49.298  step_detector   
658672  2017-07-05 13:00:52.061          audio   
1499168 2017-07-13 17:28:51.667          audio   

                                                    values      activity_type  
72433                                             ['70.0']            At home  
1236933        ['11008.0', '6452.8867', '32767.0', '35.0']                     
1257764           ['-20.56704', '-33.743465', '45.490875']             In bus  
1237425           ['-1.2023174', '0.9798073', '-9.683225']  Shopping& wearing  
892199                                       ['985.40344']               Walk  
1357285  ['-96.4684990069929', '0.9737792161950717', '-...                     
127121                                      ['120.0', '0']                     
562529                                             ['1.0']                     
658672         ['16181.0', '4240.6807', '32767.0', '48.0']                     
1499168           ['172.0', '6406.734', '32767.0', '37.0']

In [18]:
sample_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 172869 entries, 72433 to 162
Data columns (total 4 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   timestamp      172869 non-null  datetime64[ns]
 1   source         172869 non-null  object        
 2   values         172869 non-null  object        
 3   activity_type  172869 non-null  object        
dtypes: datetime64[ns](1), object(3)
memory usage: 11.6+ MB


In [26]:
filtered_df = sample_df.copy()
filtered_df.drop(filtered_df[filtered_df['activity_type'] == ''].index, inplace=True)

### **Classification**

In [27]:
filtered_df.head(10)

timestamp         source  \
72433   2017-07-03 19:56:35.104   step_counter   
1257764 2017-07-11 17:37:08.718   magnetometer   
1237425 2017-07-11 15:39:58.565        gravity   
892199  2017-07-07 17:55:46.660       pressure   
803433  2017-07-06 14:43:36.788      gyroscope   
386740  2017-07-02 17:16:06.121        gravity   
512810  2017-07-03 19:34:10.814        gravity   
503627  2017-07-03 18:47:59.665  step_detector   
1037598 2017-07-09 18:24:20.965          audio   
1152732 2017-07-10 17:06:50.117    orientation   

                                                    values      activity_type  
72433                                             ['70.0']            At home  
1257764           ['-20.56704', '-33.743465', '45.490875']             In bus  
1237425           ['-1.2023174', '0.9798073', '-9.683225']  Shopping& wearing  
892199                                       ['985.40344']               Walk  
803433     ['0.0015137931', '4.731398E-5', '4.0211622E-4']         In vehicle  
386740            ['9.212076', '-1.3891348', '-3.0624132']            Picnic   
512810           ['-7.4717717', '-5.863092', '-2.4427848']            At home  
503627                                             ['1.0']            At home  
1037598          ['1371.0', '2206.067', '32767.0', '35.0']            At home  
1152732  ['132.90780411582105', '1.1715394462065631', '...            At home

In [49]:
X = filtered_df[['source','values']]
y = filtered_df['activity_type']
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X['source'] = le.fit_transform(X['source'])
X['values'] = le.fit_transform(X['values'])
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

/opt/anaconda3/envs/tensor_flow/lib/python3.6/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/opt/anaconda3/envs/tensor_flow/lib/python3.6/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [46]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

GaussianNB()

In [38]:
y_pred  =  classifier.predict(X_test)
from sklearn.metrics import accuracy_score
ac = accuracy_score(y_test,y_pred)

In [42]:
print("Accurracy:",ac)

Accurracy: 0.29874776386404295


## **Conclusion**

In conclusion, the data seemed to have three different sources (Glassess, watch and phone). And a binding data set with user activity. It was  observed that the source glasses was identified to be incompatible with other dataset as it is, given the timeframe for the project it has been overlooked.Given time the features from the glasses would ideally be incprporated into the classifier. The dataset with smart watch and phone data were merged. The total record tally was about 1.7 million records. 

Subsequently, the data was to be labelled with activity type and timestamp as pivot. Considering the compute power of my machine, labelling million records was not viable. A sample of 0.1% was chosen at random. This sample was then labelling using a greedy loop and records with no labels were removed. With 80|20 train test split a naive bayes classifier was used to classify the dataset. This resulted with a accuracy of 29.87%. The accuracy observed was low as the features in consideration were not properly processed. Given more time and compute power, the values column in the dataset would be streamlined. The data in that column was erractic, it would ideally be extracted into sub columns with computational significance.